common/txmgr/types/head_view.go

```diff
@@ -0,0 +1,32 @@
+package types
+
+import "github.com/ethereum/go-ethereum/common"
+
+// HeadView provides minimal access to a chain's head, as needed by the TxManager.
+// This is a generic interface which ALL chains will implement.
+//
+//go:generate mockery --quiet --name HeadView --output ./mocks/ --case=underscore
+type HeadView[HEAD any] interface {
+	// BlockNumber is the head's block number
+	BlockNumber() int64
+
+	// ChainLength returns the length of the chain followed by recursively looking up parents
+	ChainLength() uint32
+
+	// EarliestInChain traverses through parents until it finds the earliest one
+	EarliestInChain() HeadView[HEAD]
+
+	// Hash is the head's block hash
+	Hash() common.Hash
+
+	// Parent is the head's parent block
+	Parent() HeadView[HEAD]
+
+	// HashAtHeight returns the hash of the block at the given height, if it is in the chain.
+	// If not in chain, returns the zero hash
+	HashAtHeight(blockNum int64) common.Hash
+
+	// GetNativeHead returns the head in the chain's native type
+	// Chain specific code can retrieve the native type via this function.
+	GetNativeHead() HEAD
```

Comment:
```suggestion
	Native() HEAD
```

Reply:
Sounds good

Reply:
I'm also not sure about this call. The chain-specific txm implements the interface to comply with the minimum required functionality based on the `HEAD` type which is known to it.

Reply:
The HeadView was a generic interface with a few functions that all chains implement. But this Native() will return the complete Head object in chain's native type. This is being passed to GasEstimators, which need the whole chain specific head object.

Reply:
method rename done

Reply:
> The HeadView was a generic interface with a few functions that all chains implement. But this Native() will return the complete Head object in chain's native type. This is being passed to GasEstimators, which need the whole chain specific head object.

What problem does this abstract wrapper solve? All the callers still have to import and reference the native type anyways.

Reply:
Components inside the core Txm don't need to know or import native types.
For example, look at the eth_confirmer.go file.
It uses the methods in the Head interface, without knowing the native type.

Reply:
> Components inside the core Txm don't need to know or import native types.

Then they should use a less complex abstraction that does not impose knowing the native type - i.e. only an interface.

> It uses the methods in the Head interface, without knowing the native type.

It _does_ know the type. It is leaking everywhere. If the generic type is not being _used_, then it should be eliminated from that code scope with a stronger abstraction that assumes less.

Reply:
I think there may be another way, which you might prefer.

I create a generic Head interface, similar to the one I currently have here.
Then, the native evmtypes.Head should implement it, and extend it with its own native head fields. Similarly, all chains will implement it, and extend with their own fields.

The txmgr will only use the generic Head interface, and pass it around. Chain-specific code can keep using native Head type.


I didn't do it this way earlier, as I didn't want to change the existing evmtypes.Head, as its being used in a lot of places.

Reply:
That makes sense to me if it can be done.